File: solutions/3113. Find the Number of Subarrays Where Boundary Elements Are Maximum/3113.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  long long numberOfSubarrays(const vector<int>& nums) {
    long long ans = 0;
    vector<pair<int, int>> stack;

    for (const int num : nums) {
      while (!stack.empty() && stack.back().first < num)
        stack.pop_back();
      if (stack.empty() || stack.back().first != num)
        stack.emplace_back(num, 0);
      ans += ++ stack.back().second;    
    }

    return ans;
  }
};

File: solutions/3113. Find the Number of Subarrays Where Boundary Elements Are Maximum/3113.rs
// Time complexity: O(n)
// Space complexity: O(n)
impl Solution {
    pub fn number_of_subarrays(nums: Vec<i32>) -> i64 {
        let mut ans: i64 = 0;
        let mut stack: Vec<(i32, i64)> = Vec::new();

        for &num in &nums {
            while !stack.is_empty() && stack.last().unwrap().0 < num {
                stack.pop();
            }
            if stack.is_empty() || stack.last().unwrap().0 != num {
                stack.push((num, 0));
            }
            ans += stack.last_mut().unwrap().1 + 1;
            stack.last_mut().unwrap().1 += 1;
        }

        ans
    }
}
